docs/configuration-reference.md
# Swoole Bundle Configuration

Documentation of available configuration parameters. See also symfony [bundle configuration](./../src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php) file or [swoole documentation](https://github.com/swoole/swoole-docs/tree/master/modules).

- [Swoole Bundle Configuration](#swoole-bundle-configuration)
  - [HTTP Server](#http-server)

## HTTP Server

*Example*:

```yaml
swoole:
    http_server:
        port: 9501
        host: 0.0.0.0
        running_mode: process
        socket_type: tcp
        ssl_enabled: false
        trusted_hosts: localhost,127.0.0.1
        trusted_proxies:
            - '*'
            - 127.0.0.1/8
            - 192.168.2./16

        # enables static file serving
        static: advanced
        # equals to:
        # ---
        # static:
        #     public_dir: '%kernel.project_dir%/public'
        #     strategy: advanced
        #     mime_types: ~
        # ---
        # strategy can be one of: (default) auto, off, advanced, default
        #   - off: turn off feature
        #   - auto: use 'advanced' when debug enabled or not production environment
        #   - advanced: use request handler class \K911\Swoole\Server\RequestHandler\AdvancedStaticFilesServer
        #   - default: use default swoole static serving (faster than advanced, but supports less content types)
        # ---
        # mime types registration by file extension for static files serving in format: 'file extension': 'mime type'
        # this only works when 'static' strategy is set to 'advanced'
        #
        #   mime_types:
        #       '*': 'text/plain' # fallback override
        #       customFileExtension: 'custom-mime/type-name'
        #       sqlite: 'application/x-sqlite3'

        # enables hot module reload using inotify
        hmr: auto
        # hmr can be one of: off, (default) auto, inotify
        #   - off: turn off feature
        #   - auto: use inotify if installed in the system
        #   - inotify: use inotify

        # enables api server on specific port
        # by default it is disabled (can be also enabled using --api flag via cli)
        api: true
        # equals to:
        # ---
        # api:
        #     enabled: true
        #     host: 0.0.0.0
        #     port: 9200

        # additional swoole symfony bundle services
        services:

            # see: \K911\Swoole\Bridge\Symfony\HttpFoundation\TrustAllProxiesRequestHandler
            trust_all_proxies_handler: true

            # see: \K911\Swoole\Bridge\Symfony\HttpFoundation\CloudFrontRequestFactory
            cloudfront_proto_header_handler: true

            # see: \K911\Swoole\Bridge\Doctrine\ORM\EntityManagerHandler
            entity_manager_handler: true
            
            # see: \K911\Swoole\Bridge\Upscale\Blackfire\WithProfiler
            blackfire_profiler: false


        # swoole http server settings
        # see https://www.swoole.co.uk/docs/modules/swoole-server/configuration
        settings:
            reactor_count: 2
            worker_count: 4
            # when not set, swoole sets these are automatically set based on count of host CPU cores
            task_worker_count: 2 # one of: positive number, "auto", or null to disable creation of task worker processes (default: null)

            log_level: auto
            # can be one of: (default) auto, debug, trace, info, notice, warning, error
            #   - auto: when debug set to debug, when not set to notice
            #   - {debug,trace,info,notice,warning,error}: see swoole configuration

            log_file: '%kernel.logs_dir%/swoole_%kernel.environment%.log'
            pid_file: /var/run/swoole_http_server.pid

            buffer_output_size: 2097152
            # in bytes, 2097152b = 2 MiB

            package_max_length: 8388608
            # in bytes, 8388608b = 8 MiB
        
            worker_max_request: 0
            # integer >= 0, indicates the number of requests after which a worker reloads automatically
            # This can be useful to limit memory leaks
            worker_max_request_grace: ~
            # 'grace period' for worker reloading. If not set, default is worker_max_request / 2. Worker reloads
            # after 'worker_max_request + rand(0,worker_max_request_grace)' requests

            cooperative_scheduling: false
            # default false. when enabled, swoole coroutine hooks for IO apis get activated
            # (https://www.swoole.co.uk/docs/modules/swoole-coroutine-enableCoroutine) and all stateful services
            # are being used in contextual way with multiple instances for each service
            # (https://www.swoole.co.uk/article/isolating-variables-with-coroutine-context)
```

CHANGELOG.md
# [0.10.0](https://github.com/k911/swoole-bundle/compare/v0.9.0...v0.10.0) (2021-05-05)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.9.0...v0.10.0)

### Features

* **response-processor:** hiding Swoole StreamedResponse support from Symfony ([#509](https://github.com/k911/swoole-bundle/issues/509)) ([ebe8077](https://github.com/k911/swoole-bundle/commit/ebe80770e9132b45f6a74ad233f8b8b707854f2e))

# [0.9.0](https://github.com/k911/swoole-bundle/compare/v0.8.3...v0.9.0) (2021-02-03)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.8.3...v0.9.0)

### Bug Fixes

* **request-factory:** Avoid accessing undefined index REQUEST_URI ([#422](https://github.com/k911/swoole-bundle/issues/422)) ([807ba9f](https://github.com/k911/swoole-bundle/commit/807ba9f0c7ceaa523af76899ce154c816ab69242))


### Features

* **response-processor:** add support for StreamedResponse ([89fc7ca](https://github.com/k911/swoole-bundle/commit/89fc7cac9864465cc707224422e81ad0d101edc4))

## [0.8.3](https://github.com/k911/swoole-bundle/compare/v0.8.2...v0.8.3) (2021-01-03)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.8.2...v0.8.3)

### Bug Fixes

* **session-storage:** Reset session storage on kernel.finish_request ([6b7a992](https://github.com/k911/swoole-bundle/commit/6b7a9923bece217e1cfa43b7fb6cd0016d8069af))
* allow defining no log file path to enable logging to stdout ([#301](https://github.com/k911/swoole-bundle/issues/301)) ([eea4a4f](https://github.com/k911/swoole-bundle/commit/eea4a4f41e6e8100affb763b18fa62403d30e705))

## [0.8.2](https://github.com/k911/swoole-bundle/compare/v0.8.1...v0.8.2) (2020-07-20)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.8.1...v0.8.2)

### Miscellaneous

* Minor fixes

## [0.8.1](https://github.com/k911/swoole-bundle/compare/v0.8.0...v0.8.1) (2020-07-14)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.8.0...v0.8.1)

### Bug Fixes

* **doctrine:** autoconfigure EntityManagerHandler only when orm is available in symfony's container ([#274](https://github.com/k911/swoole-bundle/issues/274)) ([87ede15](https://github.com/k911/swoole-bundle/commit/87ede156bc25f3b774a9913cff9db3d78d39b129))
* **http:** proper creation of $_SERVER['REQUEST_URI'] ([#269](https://github.com/k911/swoole-bundle/issues/269)) ([78bb42b](https://github.com/k911/swoole-bundle/commit/78bb42b559faae94f7eeac3aebb8886992714214))

# [0.8.0](https://github.com/k911/swoole-bundle/compare/v0.7.9...v0.8.0) (2020-06-23)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.9...v0.8.0)

### Bug Fixes

* **config:** allow configuring worker and reactor counts using ENV variables ([#244](https://github.com/k911/swoole-bundle/issues/244)) ([d6b270a](https://github.com/k911/swoole-bundle/commit/d6b270a3d3f2895d6cef55b3272f57ae30d48657))
* **profiler:** make log collection in symfony profiler to work ([#242](https://github.com/k911/swoole-bundle/issues/242)) ([50fdd6f](https://github.com/k911/swoole-bundle/commit/50fdd6fa0ab98b83c30a456405e203f6296cf2fd))


### Features

* **blackfire:** Add bridge for upscale/swoole-blackfire ([#221](https://github.com/k911/swoole-bundle/issues/221)) ([960ddb8](https://github.com/k911/swoole-bundle/commit/960ddb84004bf3b146cdddd81cc60a48c60efa0b))
* **exception-handler:** Add Symfony error/exception handler ([#228](https://github.com/k911/swoole-bundle/issues/228)) ([180d5e5](https://github.com/k911/swoole-bundle/commit/180d5e5a11b67097c66a0a2fecaf292bfa14cc4c))
* **http-server:** configurable mime types for advanced static files server ([#240](https://github.com/k911/swoole-bundle/issues/240)) ([07896a4](https://github.com/k911/swoole-bundle/commit/07896a45dccf22320cb26e0988fa4988d18cd782))

## [0.7.9](https://github.com/k911/swoole-bundle/compare/v0.7.8...v0.7.9) (2020-05-20)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.8...v0.7.9)

### Bug Fixes

* **server:** add worker_max_request and worker_max_request_grace configuration options ([#220](https://github.com/k911/swoole-bundle/issues/220)) ([69fd435](https://github.com/k911/swoole-bundle/commit/69fd435717833d224ee6087c46fd96155e362d02))

## [0.7.8](https://github.com/k911/swoole-bundle/compare/v0.7.7...v0.7.8) (2020-05-03)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.7...v0.7.8)

### Miscellaneous

* Minor fixes

## [0.7.7](https://github.com/k911/swoole-bundle/compare/v0.7.6...v0.7.7) (2020-05-01)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.6...v0.7.7)

### Miscellaneous

* Minor fixes

## [0.7.6](https://github.com/k911/swoole-bundle/compare/v0.7.5...v0.7.6) (2020-04-01)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.5...v0.7.6)

### Bug Fixes

* **http-server:** return all headers with the same name joined by comma ([#175](https://github.com/k911/swoole-bundle/issues/175)) ([1e51639](https://github.com/k911/swoole-bundle/commit/1e51639306bb8f8b4d4f636c1ef1db80a6ab9b7f))

## [0.7.5](https://github.com/k911/swoole-bundle/compare/v0.7.4...v0.7.5) (2019-12-19)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.4...v0.7.5)

### Bug Fixes

* **server:** Add missing setting for "package_max_length" ([#96](https://github.com/k911/swoole-bundle/issues/96)) ([37758f2](https://github.com/k911/swoole-bundle/commit/37758f280e853d9c40c5dfe124ff3036e66d5294))

## [0.7.4](https://github.com/k911/swoole-bundle/compare/v0.7.3...v0.7.4) (2019-12-03)

[Full changelog](https://github.com/k911/swoole-bundle/compare/v0.7.3...v0.7.4)

### Bug Fixes

* **server:** Use `SplFile::getRealPath` for `Response::sendfile` Operation for a BinaryFileReponse ([#91](https://github.com/k911/swoole-bundle/issues/91)) ([0278db7](https://github.com/k911/swoole-bundle/commit/0278db725c6b5f52bb513454a59ca16bef67f1da)), closes [#90](https://github.com/k911/swoole-bundle/issues/90)


src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php
<?php

declare(strict_types=1);

namespace K911\Swoole\Bridge\Symfony\Bundle\DependencyInjection;

use function K911\Swoole\decode_string_as_set;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;
use Symfony\Component\Config\Definition\Exception\InvalidTypeException;

final class Configuration implements ConfigurationInterface
{
    public const DEFAULT_PUBLIC_DIR = '%kernel.project_dir%/public';
    public const DEFAULT_WATCH_DIR = '%kernel.project_dir%/';

    private const CONFIG_NAME = 'swoole';

    private $builder;

    public function __construct(TreeBuilder $builder)
    {
        $this->builder = $builder;
    }

    /**
     * {@inheritdoc}
     *
     * @throws \InvalidArgumentException
     * @throws \RuntimeException
     */
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $this->builder->getRootNode()
            ->children()
                ->arrayNode('http_server')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('host')
                            ->cannotBeEmpty()
                            ->defaultValue('0.0.0.0')
                        ->end()
                        ->scalarNode('port')
                            ->cannotBeEmpty()
                            ->defaultValue(9501)
                        ->end()
                        ->arrayNode('trusted_hosts')
                            ->defaultValue([])
                            ->prototype('scalar')->end()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(fn ($v): array => decode_string_as_set($v))
                            ->end()
                        ->end()
                        ->arrayNode('trusted_proxies')
                            ->defaultValue([])
                            ->prototype('scalar')->end()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(fn ($v): array => decode_string_as_set($v))
                            ->end()
                        ->end()
                        ->enumNode('running_mode')
                            ->cannotBeEmpty()
                            ->defaultValue('process')
                            ->values(['process', 'reactor', 'thread'])
                        ->end()
                        ->enumNode('socket_type')
                            ->cannotBeEmpty()
                            ->defaultValue('tcp')
                            ->values(['tcp', 'tcp_ipv6', 'udp', 'udp_ipv6', 'unix_dgram', 'unix_stream'])
                        ->end()
                        ->booleanNode('ssl_enabled')
                            ->defaultFalse()
                            ->treatNullLike(false)
                        ->end()
                        ->arrayNode('hmr')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifTrue(function ($v): bool {
                                    return \is_string($v) || \is_bool($v) || \is_numeric($v) || null === $v;
                                })
                                ->then(function ($v): array {
                                    return [
                                        'enabled' => \in_array($v, ['auto', 'inotify', 'fsnotify'], true),
                                        'type' => \in_array($v, ['auto', 'inotify', 'fsnotify'], true) ? $v : 'auto',
                                        'watch_dir' => self::DEFAULT_WATCH_DIR,
                                        'verbose_output' => false,
                                        'tick_duration' => 5,
                                    ];
                                })
                            ->end()
                            ->children()
                                ->booleanNode('enabled')
                                    ->defaultFalse()
                                ->end()
                                ->enumNode('type')
                                    ->defaultValue('auto')
                                    ->values(['auto', 'inotify', 'fsnotify'])
                                ->end()
                                ->integerNode('tick_duration')
                                    ->min(0)
                                    ->defaultValue(5)
                                    ->max(60)
                                ->end()
                                ->booleanNode('verbose_output')
                                    ->defaultFalse()
                                ->end()
                                ->scalarNode('watch_dir')
                                    ->defaultValue(self::DEFAULT_WATCH_DIR)
                                ->end()
                            ->end()
                        ->end()
                        ->arrayNode('api')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifTrue(fn ($v): bool => \is_string($v) || \is_bool($v) || \is_numeric($v) || null === $v)
                                ->then(function ($v): array {
                                    return [
                                        'enabled' => (bool) $v,
                                        'host' => '0.0.0.0',
                                        'port' => 9200,
                                    ];
                                })
                            ->end()
                            ->children()
                                ->booleanNode('enabled')
                                    ->defaultFalse()
                                ->end()
                                ->scalarNode('host')
                                    ->cannotBeEmpty()
                                    ->defaultValue('0.0.0.0')
                                ->end()
                                ->scalarNode('port')
                                    ->cannotBeEmpty()
                                    ->defaultValue(9200)
                                ->end()
                            ->end()
                        ->end()
                        ->arrayNode('static')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(function ($v): array {
                                    return [
                                        'strategy' => $v,
                                        'public_dir' => 'off' === $v ? null : self::DEFAULT_PUBLIC_DIR,
                                        'mime_types' => [],
                                    ];
                                })
                            ->end()
                            ->children()
                                ->enumNode('strategy')
                                    ->defaultValue('auto')
                                    ->treatFalseLike('off')
                                    ->values(['off', 'default', 'advanced', 'auto'])
                                ->end()
                                ->scalarNode('public_dir')
                                    ->defaultValue(self::DEFAULT_PUBLIC_DIR)
                                ->end()
                                ->variableNode('mime_types')
                                    ->info('File extensions to mime types map.')
                                    ->defaultValue([])
                                    ->validate()
                                        ->always(function ($mimeTypes) {
                                            $validValues = [];

                                            foreach ((array) $mimeTypes as $extension => $mimeType) {
                                                $extension = \trim((string) $extension);
                                                $mimeType = \trim((string) $mimeType);

                                                if ('' === $extension || '' === $mimeType) {
                                                    throw new InvalidTypeException(\sprintf('Invalid mime type %s for file extension %s.', $mimeType, $extension));
                                                }

                                                $validValues[$extension] = $mimeType;
                                            }

                                            return $validValues;
                                        })
                                    ->end()
                                ->end() // end mime types
                            ->end()
                        ->end() // end static
                        ->arrayNode('exception_handler')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(function ($v): array {
                                    return [
                                        'type' => $v,
                                        'verbosity' => 'auto',
                                        'handler_id' => null,
                                    ];
                                })
                            ->end()
                            ->children()
                                ->enumNode('type')
                                    ->cannotBeEmpty()
                                    ->defaultValue('auto')
                                    ->treatFalseLike('auto')
                                    ->values(['json', 'production', 'symfony', 'custom', 'auto'])
                                ->end()
                                ->enumNode('verbosity')
                                    ->cannotBeEmpty()
                                    ->defaultValue('auto')
                                    ->treatFalseLike('auto')
                                    ->values(['trace', 'verbose', 'default', 'auto'])
                                ->end()
                                ->scalarNode('handler_id')
                                    ->defaultNull()
                                ->end()
                            ->end()
                        ->end() // end exception_handler
                        ->arrayNode('services')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->booleanNode('debug_handler')
                                    ->defaultNull()
                                    ->setDeprecated(
                                        'k911/swoole-bundle',
                                        '0.11',
                                        'The "%node%" option is deprecated. It is no longer needed to provide debug http kernel.'
                                    )
                                ->end()
                                ->booleanNode('trust_all_proxies_handler')
                                    ->defaultFalse()
                                    ->treatNullLike(false)
                                ->end()
                                ->booleanNode('cloudfront_proto_header_handler')
                                    ->defaultFalse()
                                    ->treatNullLike(false)
                                ->end()
                                ->booleanNode('entity_manager_handler')
                                    ->defaultNull()
                                ->end()
                                ->booleanNode('blackfire_profiler')
                                    ->defaultNull()
                                ->end()
                                ->booleanNode('session_cookie_event_listener')
                                    ->defaultFalse()
                                    ->treatNullLike(false)
                                ->end()
                            ->end()
                        ->end() // drivers
                        ->arrayNode('settings')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('log_file')
                                    // TODO: NEXT MAJOR - remove default value
                                    ->defaultValue('%kernel.logs_dir%/swoole_%kernel.environment%.log')
                                ->end()
                                ->enumNode('log_level')
                                    ->cannotBeEmpty()
                                    ->values(['auto', 'debug', 'trace', 'info', 'notice', 'warning', 'error'])
                                    ->defaultValue('auto')
                                ->end()
                                ->scalarNode('pid_file')
                                    ->defaultNull()
                                ->end()
                                ->scalarNode('buffer_output_size')
                                    ->defaultValue(2097152)
                                ->end()
                                ->scalarNode('package_max_length')
                                    ->defaultValue(8388608)
                                ->end()
                                ->scalarNode('worker_count')
                                    ->defaultValue(1)
                                ->end()
                                ->scalarNode('reactor_count')
                                    ->defaultValue(1)
                                ->end()
                                ->scalarNode('task_worker_count')
                                    ->defaultNull()
                                ->end()
                                ->integerNode('worker_max_request')
                                    ->min(0)
                                    ->defaultValue(0)
                                ->end()
                                ->scalarNode('worker_max_request_grace')
                                    ->defaultNull()
                                ->end()
                            ->end()
                        ->end() // settings
                    ->end()
                ->end() // server
            ->end()
        ;

        return $this->builder;
    }

    public static function fromTreeBuilder(): self
    {
        $treeBuilderClass = TreeBuilder::class;

        return new self(new $treeBuilderClass(self::CONFIG_NAME));
    }
}
